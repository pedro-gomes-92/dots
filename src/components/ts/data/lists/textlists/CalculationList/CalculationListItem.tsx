import React, { FunctionComponent } from 'react';
import classnames from 'classnames';

import { BaseTextListItem, BaseTextListItemProps } from '../BaseTextListItem';

export interface CalculationListItemProps extends BaseTextListItemProps {}

export const CalculationListItem: FunctionComponent<CalculationListItemProps> = ({
  className,
  ...rest
}: CalculationListItemProps): JSX.Element => (
  <BaseTextListItem className={classnames('list-calculation-item', className)} {...rest} />
);
