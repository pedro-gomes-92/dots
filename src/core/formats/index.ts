export * from './number';
export * from './duration';
export * from './percent';
export * from './currency';
export * from './date';
