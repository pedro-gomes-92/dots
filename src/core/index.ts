export * from './utilities';
export * from './formats';
