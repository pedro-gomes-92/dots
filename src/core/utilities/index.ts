export * from './object';
export * from './number';
export * from './array';
export * from './date';
export * from './duration';
export * from './currency';
