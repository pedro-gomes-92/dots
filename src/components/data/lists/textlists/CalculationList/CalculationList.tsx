import React, { FunctionComponent, useCallback } from 'react';
import classnames from 'classnames';

import { ArrayUtils } from '@dots/core';
import { BaseTextList, BaseTextListProps } from '../BaseTextList';
import { CalculationListItem } from './CalculationListItem';
import { ListDivider } from '../ListDivider';

export interface CalculationListProps extends Omit<BaseTextListProps, 'children'> {
  labels: string[];
  values?: number[];
  resultLabel?: string;
  onFormat?: (value: number) => string;
}

export const CalculationList: FunctionComponent<CalculationListProps> = ({
  className,
  values,
  labels,
  resultLabel,
  onFormat,
  size,
  ...rest
}: CalculationListProps): JSX.Element => {
  const missingValues = labels.length - values.length;
  let currValues = [...values];
  if (missingValues > 0) {
    currValues = [...currValues, ...ArrayUtils.fill(new Array(missingValues), 0)];
  } else if (missingValues < 0) {
    currValues.splice(currValues.length - Math.abs(missingValues));
  }

  const getTotal = useCallback((): number => {
    return currValues.reduce((sum, value): number => sum + value);
  }, [currValues]);

  return (
    <BaseTextList className={classnames('list-calculation', className)} {...rest}>
      <>
        {labels.map(
          (label, index): JSX.Element => (
            <CalculationListItem
              key={`key-calculation-item-${index}`}
              text={label}
              secondaryText={onFormat(currValues[index])}
              size={size}
            />
          ),
        )}
      </>
      <ListDivider />
      <CalculationListItem
        text={resultLabel}
        className="list-calculation-result"
        secondaryText={onFormat(getTotal())}
        size={size}
      />
    </BaseTextList>
  );
};

CalculationList.defaultProps = {
  resultLabel: 'Result',
  values: [],
  onFormat: (value: number): string => `${value}`,
};
