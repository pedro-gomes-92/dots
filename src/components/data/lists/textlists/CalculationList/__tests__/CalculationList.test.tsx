import React from 'react';
import { render } from '@testing-library/react';
import { FormatCurrency } from '@dots/core';

import { CalculationList } from '../CalculationList';

describe('Given <CalculationList />', (): void => {
  let container: HTMLElement;
  let getFormattedValues: (container: HTMLElement) => string[];
  let getValues: (container: HTMLElement) => number[];
  let getLabels: (container: HTMLElement) => string[];
  let values: number[];

  beforeAll((): void => {
    const itemsQuery = '.list-calculation-item';

    getLabels = (container: HTMLElement): string[] => {
      return [...container.querySelectorAll(`${itemsQuery} .is-primary`)].map((item): string => item.textContent);
    };

    getFormattedValues = (container: HTMLElement): string[] => {
      return [...container.querySelectorAll(`${itemsQuery} .is-secondary`)].map((item): string => item.textContent);
    };

    getValues = (container: HTMLElement): number[] => {
      return getFormattedValues(container).map((item): number => parseInt(item));
    };
  });

  describe('When initializes', (): void => {
    it('Then should match the snapshot', (): void => {
      container = render(<CalculationList labels={['Test']} />).container;
      expect(container).toMatchSnapshot();
    });

    describe('And uses labels', (): void => {
      beforeEach((): void => {
        container = render(<CalculationList labels={['Test 1', 'Test 2']} />).container;
        values = getValues(container);
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should have the correct number of items', (): void => {
        expect(values).toHaveLength(3);
      });

      it('And should render a text for each label', (): void => {
        expect(getLabels(container)).toEqual(['Test 1', 'Test 2', 'Result']);
      });

      it('And should have default values for each item', (): void => {
        expect(values.every((item): boolean => item === 0)).toBeTruthy();
      });

      describe('And uses values', (): void => {
        beforeEach((): void => {
          container = render(<CalculationList labels={['Test 1', 'Test 2']} values={[1, 2]} />).container;
          values = getValues(container);
        });

        it('Then should match the snapshot', (): void => {
          expect(container).toMatchSnapshot();
        });

        it('And should have the correct number of items', (): void => {
          expect(values).toHaveLength(3);
        });

        it('And should have the corresponding values', (): void => {
          expect(values).toEqual([1, 2, 3]);
        });
      });
    });

    describe('And uses values', (): void => {
      beforeEach((): void => {
        container = render(<CalculationList labels={['Test']} values={[1, 2, 3, 4, 5]} />).container;
        values = getValues(container);
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should have the correct number of items', (): void => {
        expect(values).toHaveLength(2);
      });

      it('And should have the corresponding values', (): void => {
        expect(values).toEqual([1, 1]);
      });
    });

    describe('And uses onFormat', (): void => {
      let items: string[];

      beforeEach((): void => {
        container = render(
          <CalculationList
            labels={['Test 1', 'Test 2']}
            values={[1, 2]}
            onFormat={(value): string => FormatCurrency.format(value)}
          />,
        ).container;

        items = getFormattedValues(container);
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should format each item value', (): void => {
        expect(items).toEqual(['€1.00', '€2.00', '€3.00']);
      });
    });

    describe('And uses resultLabel', (): void => {
      beforeEach((): void => {
        container = render(<CalculationList labels={['Test 1', 'Test 2']} values={[1, 2]} resultLabel="New balance" />)
          .container;
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should change the label for result', (): void => {
        expect(getLabels(container)[2]).toEqual('New balance');
      });
    });
  });
});
