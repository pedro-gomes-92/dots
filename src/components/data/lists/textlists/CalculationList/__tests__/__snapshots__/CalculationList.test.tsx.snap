// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <CalculationList /> When initializes And uses labels And uses values Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 1
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            1
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 2
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            2
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Result
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            3
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Given <CalculationList /> When initializes And uses labels Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 1
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            0
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 2
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            0
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Result
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            0
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Given <CalculationList /> When initializes And uses onFormat Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 1
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            €1.00
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 2
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            €2.00
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Result
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            €3.00
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Given <CalculationList /> When initializes And uses resultLabel Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 1
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            1
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test 2
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            2
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            New balance
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            3
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Given <CalculationList /> When initializes And uses values Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            1
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Result
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            1
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Given <CalculationList /> When initializes Then should match the snapshot 1`] = `
<div>
  <ul
    class="list list-calculation"
  >
    <li
      class="list-item list-calculation-item"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Test
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            0
          </span>
        </div>
      </div>
    </li>
    <li
      class="list-divider"
    >
      <div
        class="divider is-divider-horizontal"
      />
    </li>
    <li
      class="list-item list-calculation-item list-calculation-result"
    >
      <div
        class="layout is-2 is-mobile columns has-alignment-center"
      >
        <div
          class="layout-item column"
        >
          <span
            class="text is-primary"
          >
            Result
          </span>
        </div>
        <div
          class="is-auto layout-item column"
        >
          <span
            class="text is-secondary"
          >
            0
          </span>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
